src/graph.py

```diff
@@ -0,0 +1,66 @@
+import itertools
+
+
+class Graph():
+    def __init__(self):
+        # Create empty graph
+        self._container = {}
+
+    def nodes(self):
+        """Return a list of all nodes."""
+        return [node for node in self._container]
+
+    def edges(self):
+        """Retun a list of all edges."""
+        edges = []
+        for node in self._container:
+            for value in self._container[node]:
+                edges.append((node, value))
+        return edges
+
+    def add_node(self, n):
+        """Add node to graph."""
+        self._container.setdefault(n, set())
+
+    def add_edge(self, n1, n2):
+        """Add edge connecting node n1 to node n2.
+
+        Creates nodes n1, n2 if necessary.
+        """
+        self.add_node(n2)
+        self._container.setdefault(n1, set()).add(n2)
+
+    def del_node(self, n):
+        """Delete node."""
+        if self.has_node(n):
+            del self._container[n]
+            for node in self._container:
+                self._container[node].discard(n)
+        else:
+            raise KeyError("Node not in graph.")
+
+    def del_edge(self, n1, n2):
+        """Delete edge connecting n1 to n2."""
+        try:
+            self._container[n1].remove(n2)
+        except (KeyError, ValueError):
+            raise ValueError("Edge not in graph.")
```

Comment:
I'm okay with this, though.  Either error signals the same condition, which is that the edge isn't present.  This could be considered an example of normalizing error conditions.  It makes a more concise API, in my opinion.